url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/view/mxMultiplicity.java

```diff
@@ -0,0 +1,240 @@
+	 * for the given connection or null if the multiplicity does not apply.
+	 *  
+	 * Parameters:
+	 * 
+	 * graph - Reference to the enclosing graph instance.
+	 * edge - Cell that represents the edge to validate.
+	 * source - Cell that represents the source terminal.
+	 * target - Cell that represents the target terminal.
+	 * sourceOut - Number of outgoing edges from the source terminal.
+	 * targetIn - Number of incoming edges for the target terminal.
+	 */
+	public String check(mxGraph graph, Object edge, Object source,
+			Object target, int sourceOut, int targetIn)
+	{
+		StringBuffer error = new StringBuffer();
+
+		if ((this.source && checkTerminal(graph, source, edge))
+				|| (!this.source && checkTerminal(graph, target, edge)))
+		{
+			if (!isUnlimited())
+			{
+				int m = getMaxValue();
+
+				if (m == 0 || (this.source && sourceOut >= m)
+						|| (!this.source && targetIn >= m))
+				{
+					error.append(countError + "\n");
+				}
+			}
+
+			if (validNeighbors != null && typeError != null && validNeighbors.size() > 0)
+			{
+				boolean isValid = checkNeighbors(graph, edge, source, target);
+
+				if (!isValid)
+				{
+					error.append(typeError + "\n");
+				}
+			}
+		}
+
+		return (error.length() > 0) ? error.toString() : null;
+	}
+
+	/**
+	 * Checks the type of the given value.
+	 */
+	public boolean checkNeighbors(mxGraph graph, Object edge, Object source,
+			Object target)
+	{
+		mxIGraphModel model = graph.getModel();
+		Object sourceValue = model.getValue(source);
+		Object targetValue = model.getValue(target);
+		boolean isValid = !validNeighborsAllowed;
+		Iterator<String> it = validNeighbors.iterator();
+
+		while (it.hasNext())
+		{
+			String tmp = it.next();
+
+			if (this.source && checkType(graph, targetValue, tmp))
+			{
+				isValid = validNeighborsAllowed;
+				break;
+			}
+			else if (!this.source && checkType(graph, sourceValue, tmp))
+			{
+				isValid = validNeighborsAllowed;
+				break;
+			}
+		}
+
+		return isValid;
+	}
+
+	/**
+	 * Checks the type of the given value.
+	 */
+	public boolean checkTerminal(mxGraph graph, Object terminal, Object edge)
+	{
+		Object userObject = graph.getModel().getValue(terminal);
+
+		return checkType(graph, userObject, type, attr, value);
+	}
+
+	/**
+	 * Checks the type of the given value.
+	 */
+	public boolean checkType(mxGraph graph, Object value, String type)
+	{
+		return checkType(graph, value, type, null, null);
+	}
+
+	/**
+	 * Checks the type of the given value.
+	 */
+	public boolean checkType(mxGraph graph, Object value, String type,
+			String attr, String attrValue)
+	{
+		if (value != null)
+		{
+			if (value instanceof Element)
+			{
+				return mxUtils.isNode(value, type, attr, attrValue);
+			}
+			else
+			{
+				return value.equals(type);
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Returns true if max is "n" (unlimited).
+	 */
+	public boolean isUnlimited()
+	{
+		return max == null || max == "n";
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.